### PR TITLE
Fixed error page on comment error

### LINKED ISSUE #477: Replying to a deleted comment results in error page due to this not being possible
Reddit doesn't allow this and raises an error. We should do a few things:

- [x] Hide the reply button if the comment was deleted
- [x] Update the API to return a non-500 error
- [x] Update the frontend to handle this new error code gracefully

This error is being raised by the reddit backend when you try to reply:

```
web_1     | praw.exceptions.APIException: DELETED_COMMENT: 'that comment has been deleted' on field 'parent'
web_1     | [2018-02-07 14:32:06] ERROR 41 [django.request] exception.py:124 - [d6d603847e3a] - Internal Server Error: /api/v0/posts/e/comments/
web_1     | Traceback (most recent call last):
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 39, in inner
web_1     |     response = get_response(request)
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
web_1     |     response = self._get_response(request)
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
web_1     |     response = self.process_exception_by_middleware(e, request)
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
web_1     |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
web_1     |     return view_func(*args, **kwargs)
web_1     |   File "/usr/local/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
web_1     |     return self.dispatch(request, *args, **kwargs)
web_1     |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 477, in dispatch
web_1     |     response = self.handle_exception(exc)
web_1     |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 437, in handle_exception
web_1     |     self.raise_uncaught_exception(exc)
web_1     |   File "/usr/local/lib/python3.6/site-packages/rest_framework/views.py", line 474, in dispatch
web_1     |     response = handler(request, *args, **kwargs)
web_1     |   File "./channels/views/comments.py", line 97, in post
web_1     |     serializer.save()
web_1     |   File "/usr/local/lib/python3.6/site-packages/rest_framework/serializers.py", line 214, in save
web_1     |     self.instance = self.create(validated_data)
web_1     |   File "./channels/serializers.py", line 420, in create
web_1     |     **kwargs
web_1     |   File "./channels/api.py", line 519, in create_comment
web_1     |     return self.get_comment(comment_id).reply(text)
web_1     |   File "/usr/local/lib/python3.6/site-packages/praw/models/reddit/mixins/replyable.py", line 16, in reply
web_1     |     return self._reddit.post(API_PATH['comment'], data=data)[0]
web_1     |   File "/usr/local/lib/python3.6/site-packages/praw/reddit.py", line 411, in post
web_1     |     return self._objector.objectify(data)
web_1     |   File "/usr/local/lib/python3.6/site-packages/praw/objector.py", line 122, in objectify
web_1     |     raise APIException(*errors[0])
web_1     | praw.exceptions.APIException: DELETED_COMMENT: 'that comment has been deleted' on field 'parent'
```

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### What are the relevant tickets?
Fixes #477 

#### What's this PR do?
It hides the reply button for deleted posts, fixes the API to return a 410 status if you try to reply to one (instead of a 500), and updates the UI to show a toast message if you do so. Also has a catchall for any non-200 error so the user gets an error message and not an error page.

#### How should this be manually tested?
- Delete a comment, verify the reply button goes away
- Open the same post in two tabs:
  - delete a comment in one tab
  - without refreshing the page in the other tab, reply to that same comment and verify you get a 🍞 message

#### Screenshots (if appropriate)
selection_123

### CODE REVIEW

**Comment by aliceriot** [comment]
The reply button goes away when I refresh the page - should it disappear immediately? (after I delete a comment, I mean)

**Comment by aliceriot** [comment]
can confirm the :toast: appears as expected!

File: channels/serializers.py [hunk 1]
```diff
@@ -399,6 +400,10 @@ def validate_removed(self, value):
         """Validate that removed is a bool"""
         return {'removed': _parse_bool(value, 'removed')}
 
+    def get_deleted(self, instance):
+        """Returns True if the comment was deleted"""
+        return instance.body == "[deleted]"  # only way to tell
```
Comment (aliceriot) [inline channels/serializers.py]: :/ why reddit

File: static/js/components/CommentForms.js [hunk 2]
```diff
@@ -158,17 +159,33 @@ const mapDispatchToProps = (dispatch, ownProps) => {
...
+            })
+          )
+          cancelReply(dispatch, formKey)()
+        })
+        .catch(err => {
+          if (err.errorStatusCode === 410) {
+            // Comment was deleted
+            dispatch(
+              setSnackbarMessage({
+                message: "Comment you are replying to was deleted"
```
Comment (aliceriot) [inline static/js/components/CommentForms.js]: Might suggest 'This comment has been deleted and cannot be replied to' or something similar, although @roberthouse54 might have a better idea of what makes sense than me :smiley:

File: static/js/components/CommentTree.js [hunk 3]
```diff
@@ -120,7 +120,7 @@ export default class CommentTree extends React.Component<*, *> {
               upvote={upvote}
               downvote={downvote}
             />
-            {atMaxDepth || moderationUI
+            {atMaxDepth || moderationUI || comment.deleted
```
Comment (aliceriot) [inline static/js/components/CommentTree.js]: might be good to add a quick test to check for this, I think there's one already that checks that `moderationUI` hides this button, maybe could just make than one into a parameterized test or something.

File: channels/utils.py [hunk 4]
```diff
@@ -105,6 +105,8 @@ def translate_praw_exceptions():
             raise NotFound() from exc
         elif exc.error_type == 'SUBREDDIT_EXISTS':
             raise ConflictException() from exc
+        elif exc.error_type == 'DELETED_COMMENT':
+            raise GoneException() from exc
```
Comment (aliceriot) [inline channels/utils.py]: do you think it's worthwhile to test this? It seems like we are exercising this code via the view test you wrote, not sure if there's value in explicitly testing this `translate_praw_exceptions` function by itself.

Comment (rhysyngsun) [reply channels/utils.py]: We don't test this anywhere else either, maybe address as a separate PR so we can cover the whole func?

Comment (aliceriot) [reply channels/utils.py]: I'm not super worried about it, just wanted to raise it if we thought there was a lot of value in having unit tests for it. Maybe a tech debt issue would be the way forward there.

Comment (rhysyngsun) [reply channels/utils.py]: Add #487

**Comment by rhysyngsun** [comment]
@aliceriot feedback addressed

File: static/js/components/CommentTree_test.js [hunk 5]
```diff
@@ -125,11 +125,17 @@ describe("CommentTree", () => {
...
-    wrapper.find(".comment-action-button").at(0).simulate("click")
+    wrapper.find(".comment-action-button.reply-button").at(0).simulate("click")
     assert.ok(beginEditingStub.called)
     assert.ok(beginEditingStub.calledWith(replyToCommentKey(comments[0])))
   })
 
+  it('should not include a "reply" button for deleted posts', () => {
+    comments[0].deleted = true
+    const wrapper = renderCommentTree()
+    assert.ok(wrapper.find('.comment').at(0).find(".comment-action-button.reply-button").exists())
```
Comment (aliceriot) [inline static/js/components/CommentTree_test.js]: shouldn't this be `assert.isNotOk`?

Comment (rhysyngsun) [reply static/js/components/CommentTree_test.js]: Err...yeah... 🤔

File: static/js/components/CommentTree_test.js [hunk 6]
```diff
@@ -125,11 +125,23 @@ describe("CommentTree", () => {
...
+    comments[0].deleted = true
+    const wrapper = renderCommentTree()
+    assert.ok(
+      wrapper
+        .find(".comment")
+        .at(0)
+        .find(".comment-action-button.reply-button")
+        .exists()
+    )
+  })
```
Comment (aliceriot) [inline static/js/components/CommentTree_test.js]: the issue with this is that comments[0].replies has undeleted comments in it, so there will still be a .reply-button somewhere in the tree. I think just setting comments[0].replies = [] should make the test fail (as it should!)

Comment (rhysyngsun) [reply static/js/components/CommentTree_test.js]: I fixed it instead by altering the selector.

Comment (rhysyngsun) [reply static/js/components/CommentTree_test.js]: I thought I had pushed this but apparently didn't since I rebased and didn't force push it.

**Comment by aliceriot** [comment]
(assuming tests pass!)

**Pull request merged**